File: apps/baseapp.py
#!/usr/bin/python
import sys
import os.path
import time
from cbcomms import *

# Sockets are named in/out as seen by the app
theapp = os.path.basename(sys.argv[0])
print "Name:"+theapp

nchannels = openAppChannels(theapp)
for ch in range(0, nchannels):
    initClientChannel(ch)

while True:
    for ch in range(0, nchannels):
        if isInputChannel(ch):
            readChannel(ch)
            buffer = getChannelData(ch)
            print "App received data from"+getChannelName(ch)+" ["+buffer+"]"

    for ch in range(0, nchannels):
        if isOutputChannel(ch):
            print "App sent data to "+getChannelName(ch)+" ["+buffer+"]"
            setChannelData(ch, buffer)
            writeChannel(ch)

    time.sleep(1)

for ch in range(0, nchannels):
    closeChannel(ch)

File: apps/baseapp.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <unistd.h>

#include "CB_commslib.h"

int main(int argc, char *argv[])
{
    CB_ls("/lxc", "");
    CB_ls("/lxc", basename(argv[0]));

    // Sockets are named in/out as seen by the app
    size_t nchannels = openAppChannels(basename(argv[0]));
    for (size_t ch = 0; ch < nchannels; ++ch)
        initClientChannel(ch);

    char buffer[1024];
    while (1)
    {
        for (size_t ch = 0; ch < nchannels; ++ch)
            if (isInputChannel(ch))
            {
                readChannel(ch);
                printf("App received data from %s [%s]\n",
                       getChannelName(ch),
                       getChannelData(ch));
                strncpy(buffer, getChannelData(ch), 1023);
            }

        for (size_t ch = 0; ch < nchannels; ++ch)
            if (isOutputChannel(ch))
            {
                printf("App sent data to %s [%s]\n",
                       getChannelName(ch),
                       buffer);
                setChannelData(ch, buffer);
                writeChannel(ch);
            }
        sleep(1);
    }

    for (size_t ch = 0; ch < nchannels; ++ch)
        closeChannel(ch);

    return 0;
}
